### PR TITLE
Driver for Invensense sensors ICM20648, ICM20948 and ICM20649

### PR DESCRIPTION
**Pull request by bugobliterator**

### CODE REVIEW

**Comment by proficnc** [comment]
Thanks Sid!!!

**Comment by tridge** [comment]
If the 20948 is on I2C, then does this change force us to init the IMU on I2C as well? Previously it instantiated with just I2C compass, in i2c pass-thru mode.
I'm just concerned we'll hit the IMU limit of 3 and lose 20948 on i2c in a GPS

**Comment by proficnc** [comment]
In a cube Yellow or orange, the ICM20948 is on SPI1
On som GPS’s an ICM20948 is being used as a magnetometer on I2C, but as you say, it is read as a bypass mode, directly to the AK09917 

So the behaviour needs to be defined by bus type I assume.

At no point do we attempt to use I2C for an IMU.

**Comment by magicrub** [comment]
This feels like we need a HAL_CPU_CLASS_200 that could support more <library>_MAX_INSTANCES. Oh look, someone had the same idea! https://github.com/ArduPilot/ardupilot/pull/10544

**Comment by tridge** [comment]
@proficnc, yep, I presumed it was on SPI, I just wondered if the old use case on I2C still worked. The current code goes via I2C directly, whereas the new code goes via the bus adapter code in the IMU. I am concerned this may break users of the GPS, and just want to make sure Sid has tested this case (with the IMU not instantiated)

File: libraries/AP_Compass/AP_Compass_AK09916.cpp [hunk 1]
```diff
@@ -66,175 +93,347 @@ AP_Compass_Backend *AP_Compass_AK09916::probe(AP_HAL::OwnPtr<AP_HAL::I2CDevice>
...
+    hal.scheduler->delay(1);
+
+    // now check if a AK09916 shows up on the bus. If it does then
+    // we have both a real AK09916 and a ICM20948 with an embedded
+    // AK09916. In that case we will fail the driver load and use
+    // the main AK09916 driver
+    if (dev->read_registers(REG_COMPANY_ID, (uint8_t *)&whoami, 2)) {
+        // a device is replying on the AK09916 I2C address, don't
+        // load the ICM20948
+        gcs().send_text(MAV_SEVERITY_INFO, "ICM20948: AK09916 bus conflict\n");
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_AK09916.cpp]: shouldn't this do a "goto fail;" ?

Comment (tridge) [reply libraries/AP_Compass/AP_Compass_AK09916.cpp]: also, shouldn't we check the whoami value?

File: libraries/AP_Compass/AP_Compass_AK09916.cpp [hunk 2]
```diff
@@ -66,175 +93,347 @@ AP_Compass_Backend *AP_Compass_AK09916::probe(AP_HAL::OwnPtr<AP_HAL::I2CDevice>
...
     }
 
-    if (bus_type == AK09916_ICM20948_I2C && dev_icm) {
-        uint8_t rval;
-        if (!dev_icm->read_registers(REG_ICM_WHOAMI, &rval, 1) ||
-            rval != ICM_WHOAMI_VAL) {
-            // not an ICM_WHOAMI
-            goto fail;
-        }
-        uint8_t retries = 5;
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_AK09916.cpp]: we seem to have lost the retry loop here?

File: libraries/AP_Compass/AP_Compass_AK09916.cpp [hunk 3]
```diff
@@ -66,175 +93,347 @@ AP_Compass_Backend *AP_Compass_AK09916::probe(AP_HAL::OwnPtr<AP_HAL::I2CDevice>
...
         goto fail;
     }
 
-    dev->setup_checked_registers(2);
+    if (!_bus->start_measurements()) {
+        gcs().send_text(MAV_SEVERITY_INFO,"AK09916: Could not start measurements\n");
+        goto fail;
+    }
 
-    dev->write_register(REG_CNTL2, 0x08, true); // continuous 100Hz
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_AK09916.cpp]: are we no longer using continuous mode?

File: libraries/AP_InertialSensor/AP_InertialSensor_Invensensev2.h [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    bool _check_raw_temp(int16_t t2);
+
+    int16_t _raw_temp;
+    
+    // instance numbers of accel and gyro data
+    uint8_t _gyro_instance;
+    uint8_t _accel_instance;
+
+    float temp_sensitivity = 1.0/340; // degC/LSB
+    float temp_zero = 36.53; // degC
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_Invensensev2.h]: did you check that temp_zero and sensitivity is the same for all these sensors? (each sensor seems to have different values for the old ones)

Comment (bugobliterator) [reply libraries/AP_InertialSensor/AP_InertialSensor_Invensensev2.h]: Yes, they are same for all, the actual value is being set here:
https://github.com/ArduPilot/ardupilot/blob/9f50feadfed3e0066e5c5fcc0db147fcb6abcac4/libraries/AP_InertialSensor/AP_InertialSensor_Invensensev2.cpp#L206
I will update the code to be done for default values instead.

File: libraries/AP_Compass/AP_Compass_AK09916.cpp (hunk 3 above)
Comment (bugobliterator) [reply libraries/AP_Compass/AP_Compass_AK09916.cpp]: Yes we are, its just mode setup is done from:
https://github.com/ArduPilot/ardupilot/blob/9f50feadfed3e0066e5c5fcc0db147fcb6abcac4/libraries/AP_Compass/AP_Compass_AK09916.cpp#L308-L310

**Comment by tridge** [comment]
thanks Sid! This can be merged when CI passes

**Pull request merged**